gbn/timeout_manager.go

```diff
@@ -0,0 +1,327 @@
+	// responses received since last updating the resend timeout.
+	responseCounter int
+
+	// timeoutUpdateFrequency represents the frequency of how many
+	// corresponding responses we need to receive until the resend timeout
+	// will be updated.
+	timeoutUpdateFrequency int
+
+	sentTimes   map[uint8]time.Time
+	sentTimesMu sync.Mutex
+}
+
+// NewTimeOutManager creates a new timeout manager.
+func NewTimeOutManager(isServer bool) *TimeoutManager {
+	return &TimeoutManager{
+		isServer:               isServer,
+		resendTimeout:          defaultResendTimeout,
+		handshakeTimeout:       defaultHandshakeTimeout,
+		useStaticTimeout:       false,
+		resendMultiplier:       defaultResendMultiplier,
+		finSendTimeout:         defaultFinSendTimeout,
+		recvTimeout:            DefaultRecvTimeout,
+		sendTimeout:            DefaultSendTimeout,
+		sentTimes:              make(map[uint8]time.Time),
+		timeoutUpdateFrequency: defaultTimeoutUpdateFrequency,
+	}
+}
+
+// Sent should be called when a message is sent by the connection. The resent
+// parameter should be set to true if the message is a resent message.
+func (m *TimeoutManager) Sent(msg Message, resent bool) {
+	if m.useStaticTimeout {
+		return
+	}
+
+	// We will dynamically update the resend timeout throughout the lifetime
+	// of the connection, to ensure that it reflects the current response
+	// time. Therefore we'll keep track of when we sent a package, and when
+	// we receive the corresponding response.
+	// If we're resending a message, we can't know if a corresponding
+	// response is the response to the resent message, or the original
+	// message. Therefore, we never update the resend timeout after
+	// resending a message.
+	switch msg := msg.(type) {
+	case *PacketSYN:
+		if resent {
+			// If we've resent the SYN, we'll reset the
+			// latestSentSYNTime to the zero value, to ensure
+			// that we don't update the resend timeout based on
+			// the corresponding response, as we can't know if
+			// the response is for the resent SYN or the original
+			// SYN.
+			m.latestSentSYNTime = time.Time{}
+		} else {
+			m.latestSentSYNTime = time.Now()
+		}
+
+	case *PacketData:
+		m.sentTimesMu.Lock()
+		defer m.sentTimesMu.Unlock()
+
+		if resent {
+			// If we're resending a data packet, we'll delete the
+			// sent time for the sequence, to ensure that we won't
+			// update the resend timeout when we receive the
+			// corresponding response.
+			delete(m.sentTimes, msg.Seq)
+		} else {
+			m.sentTimes[msg.Seq] = time.Now()
+		}
+	}
+}
+
+// Received is should be called when a message is received by the connection.
+func (m *TimeoutManager) Received(msg Message) {
+	if m.useStaticTimeout {
+		return
+	}
+
+	switch msg := msg.(type) {
+	case *PacketSYN, *PacketSYNACK:
+		if !m.isServer && !m.latestSentSYNTime.IsZero() {
+			m.updateResendTimeout(time.Since(m.latestSentSYNTime))
+		}
+
+	case *PacketACK:
+		m.sentTimesMu.Lock()
+		defer m.sentTimesMu.Unlock()
```

Comment:
Could we just do the `sentTime, ok := m.sentTimes[msg.Seq]` part here then unlock `m.sentTimesMu`? I'm always careful to not call into methods (like `m.updateResendTimeout`) that potentially locks things while holding a lock within the caller. That can very easily lead to deadlocks if not careful. So this pattern would be easier to reason about.

Reply:
Ah, yeah good idea :)! I'be updated this with the latest push :)!